### ISSUE TITLE
feature request: add afterRender event for knockout components

### ISSUE THREAD

**Issue by SuperCuke** [knockout/knockout#1533]
Hello,

I propose to add an afterRender event to components binding - the event to be fired after viewModel and template are loaded and bound. Event will be useful for example to make area visible after component is loaded.

Best regards

**Comment by yohanliyanage** [comment]
+1

Having this event would be useful IMO.

**Comment by SteveSanderson** [comment]
Yes, I see this would be useful. To make the behaviour more precisely self-descriptive, I suggest the callback would be called something like `componentWasDisplayed`, an optional method on your component viewmodel that KO invokes only once when the template is first bound.

It's important to note that this would not have anything to do with child components, as it's much less clear what it means to say the child components are loaded or displayed (the set of child components can change over time, and in general considering that component loading is asynchronous, there's no way [even theoretically](http://en.wikipedia.org/wiki/Halting_problem) to know when you've finished adding child components, unless someone can give a more useful definition for "child components have finished loading" than I can think of presently).

**Comment by brianmhunt** [comment]
It would be possible, though arguably not practical, to have a queue (or `Set`) of components, with a component being added upon construction, removed just after its bindings are applied (or some hook that defaults to binding, but may be extended to cover other user-land asynchronous elements).

If components are the only asynchronous part of the process, then watching for the queue being empty *should* cover the component binding process.

Not only is this probably not performant, but it is also very "rough", in the sense that there is no way to easily track the component hierarchy completion – one can only discern whether some component, somewhere, is between starting and binding. Building a hierarchy would likely add substantial performance overhead.

In the ordinary course, this may differ from the halting problem because the component lifecycle has a definite start (construction) and end (binding). It is correct to say that where there is an aberration like a browser error, there may be no end – which is the halting problem. (For a Javascript error one could set up a timeout to check after a reasonable period for the components that did not complete– but that is way outside what the library ought to do, IMHO.)

Given how "rough" the apparent solution is, and the potential issues with performance, my sense is that `afterRender` is not a feasible built-in for components. However for an actionable item, perhaps hooks that would permit the user to implement something along the lines of the above – e.g. hooks such as `ko.components.beforeViewConstruction` and `ko.components.afterComponentBinding` (which I hope are self-descriptive).

**Comment by bjornalm** [comment]
+1

**Comment by matthewma** [comment]
I could be wrong, I think Durandal already implemented a kind of component that could track rendered state even with nested components. IMHO, this is a common problem, even with Web Component standard, this could be an issue without proper life-cycle notification.

**Comment by kjeske** [comment]
I'm using for that purpose my custom binding. Details here: http://codepen.io/kjeske/pen/vNOjwp/ But of course it would be better to have access to this event inside the createViewModel function using the factory pattern.

**Comment by chinamerp** [comment]
+1

**Comment by herbertpimentel** [comment]
+1 it will be so amazing

**Comment by Elephant-Vessel** [comment]
+1 This is needed

**Comment by miketalley** [comment]
+1 super useful

**Comment by brianmhunt** [comment]
Another thought here is to add a listener for the [transitionend](https://developer.mozilla.org/en-US/docs/Web/Events/transitionend) event, since that bubbles.

This would allow handling of animation transitions, too.

So one could have something like:

```html
<component-name data-bind='afterTransitionEnd: component_was_rendered'>

</component>
```

for some `afterTransitionEnd` binding that listens for the event and calls a given function (in this case `component_was_rendered` when complete.

I feel like this allows for a neat separation of state and logic with JS/CSS.

Unfortunately it is IE10+, as far as I can see.

**Comment by krishnaglick** [comment]
I threw up a PR for this.
https://github.com/knockout/knockout/pull/1944

**Comment by chrisknoll** [comment]
Other than the debug.js file being put into #1994 PR, the change looks pretty straight forward.

**Comment by krishnaglick** [comment]
Iirc the auto-build did that, I checked my local copy and I didn't have it.

**Comment by AlexGodofsky** [comment]
I think the afterRender function needs to receive the constructed viewModel (and possibly the element) as an argument. Otherwise it's actually pretty tricky to get access to either.

**Comment by chrisknoll** [comment]
How is this problem unlike the observable dependency tracking we already have in knockout?  When a top level container starts rendering, it opens up a dependency tracker. Each component that gets created attaches to the open tracker, such that as child components open their own contexts.  As the leaf children components finish their rendering, they notify their dependent trackers of completion.  When the entire chain is 'resolved' a callback can be fired.  

Isn't this pretty much the same mechanic around computed observables and how they set up dependency tracking based on what observables are invoked during evaluation?

-Chris

**Comment by brianmhunt** [comment]
@chrisknoll While I haven't delved into it, it feels like the same mechanic to me, too.

The only question, in my mind, is when components are "complete" – it may be we want them to self-report, or simply after they are rendered (subject to any sub-components).

**Comment by chrisknoll** [comment]
I thought the flow would work similarly to the template binding:
1. Begin tracking
2. Instantiate component
3. get component template
4. apply bindings to template (this may create other components which will begin their own tracking, but establish a dependency on the tracking from 1.
5. resolve tracking from 1 (this may remain and open tracking due to child components)

The more I think about this, the more I wonder if this isn't something that shouldn't just apply to any data binding?  Somewhere int he bowels of knockout the dom manipulation knows where to begin..this origin triggers a 'root' render context.  Any children that are rerendered jump on the tracker of this root context...and as each one completes, it notifies their link int he chain. Once the chain is complete resolved, the root can fire a notification. Up to the client code to register a handler for that notification....but I feel like that's the essence of the mechanism.

To answer your question, a component is complete once all it's direct and descendant bindings have been applied.  This could be for-each, template, basically anything that modifies the DOM (which is essentially any data-bind, yes?)

-Chris

**Comment by brianmhunt** [comment]
@chrisknoll My thoughts exactly – a generic relationship is on my mental TODO list for stateful bindings — https://github.com/brianmhunt/knockout/pull/1

**Comment by simonmurdock** [comment]
I'd really like to see this functionality

**Comment by smashraid** [comment]
Any updated please. I'm waitkng for this features.

**Comment by lzl124631x** [comment]
+1

**Comment by joeheyming** [comment]
All you need to do is put an afterRender function on the component

function MyAwesomeComponent(params) {}

MyAwesomeComponent.prototype.afterRender = function() {
}

Then call this function after applyBindingsToDescendants finishes

// deep inside knockout code near the component binding...
ko.applyBindingsToDescendants(childBindingContext, element);
componentViewModel.afterRender();

The question is, do we need anything like the $context, or the component element?

**Comment by AlexGodofsky** [comment]
I think it ought to work the same as the template version of `afterRender`, both for consistency's sake and because all the same considerations apply. A component is basically just a template with a built-in viewmodel after all.

The template version receives an array of all its DOM nodes as `elements`. Since unlike templates a component is guaranteed to have a single root node, I'd just pass it that node as `element`.

I think receiving `$context` is a bit odd though - if your component is consuming the entire `$context` object then it's not really reusable.

**Comment by joeheyming** [comment]
+1 for consistency

**Comment by kori0129** [comment]
+1 
Using a custom binding to implement this for components, while it is already supported by the framework for templates is messy.

**Comment by caseyWebb** [comment]
I'm late to this party, but perhaps this could help someone:

```html
<my-component params="afterRender: someFunc">
```

```javascript
'use strict'

class ViewModel {
  constructor(params) {
  	params.afterRender()
  }
}

ko.components.register('my-component', {
  template: 'I\'m not empty',
  viewModel: ViewModel
})
```

https://jsfiddle.net/CaseyWebb/qucy9q8h/1/

**Comment by lzl124631x** [comment]
It has been two years since this issue being opened. Knockout is falling behind. Maybe time to consider migrating to React / Angular. 😢

**Comment by kimgronqvist** [comment]
This might have been mentioned before, but I use a component loader as a workaround:

```js
const componentLoader = {
    loadComponent(name, componentConfig, callback) {
        let ViewModel = componentConfig.viewModel;

        componentConfig.viewModel = {
            createViewModel(params, componentInfo) {
               const element = componentInfo.element;

               const vm = new ViewModel(params);

               if (vm.afterRender) {
                  $(element).append('<!-- ko template: { afterRender: afterRender.bind($data) } --><!-- /ko -->');
               }

               return vm;
           }
       };
    
       ko.components.defaultLoader.loadComponent(name, componentConfig, callback);
    }
};

class Component {
   constructor(params) {}
   afterRender() {
      console.log(this);
   }
}
```

**Comment by lzl124631x** [comment]
@kimgronqvist I've tried wrapping `template & afterRender` around component but it didn't work. The outer component's afterRender fired before the inner component's afterRender. [DEMO on CodePen](http://codepen.io/lzl124631x/pen/GNpwrz?editors=1111)

I tried your `componentLoader`, same problem 😢  [DEMO on CodePen](http://codepen.io/lzl124631x/pen/YpyRQJ?editors=1011)

**Comment by kimgronqvist** [comment]
@lzl124631x That's becuase Knockout components are async by default. You need to define the inner component with `synchronous: true`.

**Comment by lzl124631x** [comment]
@kimgronqvist Forget that. Worked [DEMO on CodePen](http://codepen.io/lzl124631x/pen/rWxZMJ?editors=1111). But the doc warns me that

> Then it might load asynchronously on first use, followed by synchronously on all subsequent uses.

So, does this mean that the component loading sequence cannot be guaranteed on the first use?

**Comment by mbest** [comment]
It will only be asynchronous if you're using the `require` option.

**Comment by mbest** [comment]
From the amount of +1 and discussion, this looks important to get into the next version. Let's figure out how specifically this should work.

**Comment by jlaustill** [comment]
+1

**Comment by codymullins** [comment]
@mbest what do you think of the solution @krishnaglick submitted above? This would be useful to have (I see it as similar to Angular's onInit callback. Will be helpful to have any API calls only done when it's specifically needed (actually initialized).

**Comment by mbest** [comment]
Assuming that we'd want the *after-render* of an outer component to trigger after an inner component, we'll need a technique that will support the async rendering of components. In any case, here are some possible choices:

1. after-render is triggered synchronously (like #1944 and other examples here). So an outer component will be "rendered" before inner components unless those components use the `synchronous` option.
2. after-render is triggered asynchronously after all inner components are rendered. This might mean a significant delay if there are lots of nested components that are each loaded over the network.
3. after-render is triggered asynchronously after all pre-loaded inner components are rendered. This is like option 1 without having to use the `synchronous` option.

**Comment by mbest** [comment]
Another choice is how to specify the *after-render* callback function. Here are some choices (maybe others):

1. An option in the component config. This can get tricky if a component is using `require`.
2. A specifically named function in the viewmodel. This is the same as the current technique for specifying a `dispose` callback. But adding a new special method could be considered a breaking change.
3. An option on the `component` binding. How would this work when using a custom element?
4. Using some sort of callback registration within the component viewmodel constructor.

**Comment by joeheyming** [comment]
hmmm, 

Re: where the function goes: I like option 2 for a named function on the viewmodel, like dispose
Maybe you could establish precedence and look in the viewmodel, or look in the params.

Regarding: when to fire the callback: how does template afterRender work?  Could component.afterRender behave the same?

**Comment by codymullins** [comment]
@mbest as far as how it's triggered, what if we got the both of best worlds and had two after-render type callbacks? One like a 'pre after-render' and another a 'post after-render'? I could see value in both, the first would be useful for API calls and things that don't really need to wait for the nested components, and the second I see value in being able to say "hey, this whole thing loaded - now show it" or something to that effect.

As for how to specify it, I'm a big fan of #2 in your list. To help prevent a breaking change (if it matters to you, it doesn't so much for me), it could always be an opt-in feature like how it was done with the deferred updates.

**Comment by chrisknoll** [comment]
@mbest: you raised a question about in a custom element?  Do you mean something like:

```
<my-component params="someModelProperty" afterRender="someCallback"></my-component>
```

and the normal component binding would be like:

```
<div data-bind='component: {
    name: "my-component",
    params: someModelProperty,
    afterRender: someCallback
}'></div>
```

Since we're just adding something new I don't see this as a breaking change, is it?

**Comment by codymullins** [comment]
@chrisknoll if we added it as a special method in the viewModel, we could leave it up to whoever is using it to pass in their own callback as a parameter anyways. 

i.e.

```
function TodoViewModel(params){
   function $afterRender(){
      console.log('component was rendered!');
      params.myCallback();
   }
}
```
<my-component params="someModelProperty: 5, someCallback: myMethod"></my-component>

**Comment by AlexGodofsky** [comment]
I think afterRender should be triggered synchronously like in #1944 and like how template afterRender currently works. Hijacking template afterRender is the current workaround and seems to work fine.

My solution to the problem of running code after the internal components are done loading is to use the knockout task queue. I think this is simpler than adding extra options or parameters.

(Possibly relevant: in my particular use case, I'm mixing knockout and jQuery-mobile (warning: this is actually very frustrating), and so I need to trigger jqm's "refresh" or "create" method on a given page (or similar container) element after all the components on the page have been rendered. I maintain a "redraw queue" and each component's afterRender adds its container element to that queue (if the element isn't already there), and there is a knockout task scheduled to empty the queue. This keeps the number of calls to refresh/create down for highly nested components.)

**Comment by mbest** [comment]
I've uploaded two implementation of `afterRender`. Both use a function on the component viewModel for the callback. The difference is in how and when the callback is triggered.

* https://github.com/knockout/knockout/compare/1533-component-afterrender-trigger-option2
* https://github.com/knockout/knockout/compare/1533-component-afterrender-trigger-option3

Neither of these are final. They are just to show how the two options could be implemented. I added a small set of tests that pass for both versions, obviously not testing the difference between them.

**Comment by rabeesh** [comment]
+1 for this support

**Comment by jacobslusser** [comment]
If it helps anyone else, I've dealt with this issue also. I've come to the conclusion that the `component` binding serves a different purpose than the `template` binding and if what you want is the `afterRender` function then you should use the `template` binding. At first I was disheartened by that realization because "components" are the greatest thing since sliced bread... except, they're not. What made components cool for me was the `require` functionality built-in that worked nicely with my RequireJS setup and could dynamically load views/viewmodels on demand.

Did you know that it's actually pretty easy to use `require` with templates also?

Rather than repeat myself in two different threads, I'll just refer you to my [comments](https://github.com/knockout/knockout/issues/1475#issuecomment-266872598) in issue #1475 and the Gist I put together demonstrating how to dynamically load templates: [https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404](https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404).

Using a simple RequireJS plugin, it can be as easy as adding a dependency to `'component!<name>'` and get either the component OR a template/model.

I have used this technique on large RequireJS/Knockout projects and it works great! Both the `component` and the `template` binding are supported. In practice we find that we use the `template` binding more often. It is just superior in almost every way to the `component` binding and when the hard part of loading templates dynamically is addressed, using the `template` binding is just as easy as the `component` binding but far more flexible.

Use the `component` and `template` bindings appropriately. When done right, I would take RequireJS/Knockout over AngularJS any day of the week and twice on Sundays.

**Comment by smashraid** [comment]
Any update on this por possible workaround. O have issues executing jQuery plugins

**Comment by jonbnewman** [comment]
Throwing in the solution we use in [FootworkJS](http://footworkjs.com)...essentially it is solved via a custom $lifecycle binding that is wrapped around all rendered components which handles the afterRender capability via the *update* callback on the binding.

The $lifecycle binding:
https://github.com/footworkjs/footwork/blob/master/source/binding/lifecycle-binding.js

The custom component loader that wraps all components with the binding:
https://github.com/footworkjs/footwork/blob/master/source/component/component-lifecycle-loader.js

How it ties into the view model:
http://docs.footworkjs.com/release/2.0.1/viewModel-creation/#afterrender-callback

Just throwing this out there as an example, in-use solution...

**Comment by codymullins** [comment]
@mbest have you had any more thoughts on this?

**Comment by chrisknoll** [comment]
@mbest, I haven't forgotten about this, but am inclined to trust that you know the best forward-thinking approach to implementing the feature, so if you have something in hand that no one has objections to, I'm totally behind it.

**Comment by codymullins** [comment]
Yeah, I'd love to weigh in more on implementation too but I'm unfortunately still not proficient in understanding the entire implementation of Knockout - I trust you here too.

**Comment by caseyWebb** [comment]
If you're building a SPA, [ko-component-router](https://github.com/Profiscience/ko-component-router) supports beforeRender, afterRender, beforeDispose, and afterDispose lifecycle hooks.

DISCLAIMER: It's my package.

**Comment by codymullins** [comment]
@caseyWebb still waiting on official IE9 support ;)

**Comment by mbest** [comment]
I'm coming back to this since I want to include this feature in some sort in the next release. My idea now is to use a `Promise` that will resolve once a component, and its descendants, are displayed. This `Promise` will  be available as part of the `componentInfo` parameter passed to the `createViewModel` function.

**Comment by chrisknoll** [comment]
Promises make sense when the code looks like you are initiating something and then you are promised a result....making an ajax request results in a promise...in this case, however, you don't exactly initiate a 'initiate component rendering', that happens a bit under the covers.  Instead, this feels more like an event hook or some type of event that you want to subscribe to or register a callback handler for...similar to what a promise is for, of course, but different in that the 'initiating the action' doesn't get started and you don't get a handle on that to-be-completed action (int he form of a promise).  Maybe defining a standard method (like we have for disposal) on the component model that you can specify that will be called (if specified) once the component (and children) are done.

I like the form:

```
<div data-bind="component: {name: someComponentName, onRenderComplete: someCallback}"></div>

<someComponentName params="someParams" onRenderComplete="someCallback"></someComponentName>
```

**Comment by mbest** [comment]
Hey @chrisknoll, thanks for the feedback. I'm wary about linking this feature to a method on the viewmodel, as it would introduce binding-level information to the component viewmodel, which we've already specifically tried to avoid by distinguishing between a viewmodel construction and a `createViewModel` method. This leaves one of the [other three options](https://github.com/knockout/knockout/issues/1533#issuecomment-261668104). I like the callback registration option, using a property on the `componentInfo` object.

`Promise` provides a convenient way to register a callback that needs to be called only once when something is complete. It seems that this is the way JavaScript and the DOM are moving. For example, see [`document.fonts.ready`](https://drafts.csswg.org/css-font-loading/#font-load-event-examples).

**Comment by chrisknoll** [comment]
Hey, @mbest , so I think your document.fonts.ready nails it:  it's a promsie that isn't actually created by the 'user' (ie: the dev) instead it's just known to exist and you can do a .then() on it.  Maybe that's exacly what you should do on the component just not sure exactly which option this falls under, but maybe using the syntax options i desribed above:

````
<someComponentName params="someParams" onReady="someCallback"></someComponentName>
```
and in the guts of knockout: every initlaized component gets a 'ready' promise attached to it, and the aboe binding will invoke 'someCallback' when the ready promise is resolved.  

I know I am re-hashing up what I just said before, but my confusion is where you look for the 'ready promise' in the API.  In the example with the fonts, some underlying framework (ie: the browser's dom) set up the ready promise and attached it to document.fonts.ready.  It's like a standard pre-defined place to find it.  In the case of components, I'm not sure how that looks.  I think I've tried to get a reference to a component, usually it means I pass in a observiable to the component and then in the constructor function I assign 'this' back to the passed in observable..and it felt dirty...and also 'roll-your-own'.  

So my current feeling is I don't quite see how it 'works' maybe a code example of how it would work would help.  my discionnect is where the promise is created, and then how it's referenced later so you can .then() to it.

**Comment by vvs** [comment]
While we are at it, maybe a proper lifecycle for the components should be defined, not only a single onReady/afterRender?

Take a look at Vue.js take on component's lifecycle:
https://vuejs.org/v2/guide/instance.html#Lifecycle-Diagram

There's also Oracle JET's ojModule take (search for "ViewModel's Lifecycle"):
https://docs.oracle.com/middleware/jet210/jet/reference-jet/ojModule.html

**Comment by IanYates** [comment]
@chrisknoll - I think what you're asking for is pretty much what @mbest is suggesting.
The difference is that, in your proposal, the callback would be in the parent component whereas @mbest is having the child component (the one you're waiting to render) be the thing to receive the callback.
Note that when I say callback I really mean promise - since, as you described above, a promise is really a nice way of having a handle to which you can attach a callback for completion.
The promise approach certainly makes sense IMHO.

The differing approaches do raise an interesting point - which "component" is interested in the completion of rendering.  

We have
`<componentA></componentA>`

In that component some {{#if}} binding's condition becomes true and we end up needing to show componentB  (which may well have componentC, D and E within it) - that componentB is going to be inside of componentA, like this

```
<componentA>
    <componentB>
    ...
    </componentB>
</componentA>
```

With @mbest's suggestion, which is what I thought too, it's the `createViewModel` function of componentB that will get the notification that componentB, and its children (C, D & E) have all rendered.  

ComponentA is unaware.  It's componentInfo and `createViewModel` function ran well before our code even decided that componentB should be displayed.
However you do raise a good point - what if it's useful for componentA to know that componentB has finished rendering?

Is there merit in handling the latter approach in addition to the componentInfo/createViewModel approach suggested by @mbest?  The latter approach can be done manually via a component parameter provided by A that B accepts - it's just whether KO should take over this responsibility or if it should be explicit in component B's contract.

**Comment by chrisknoll** [comment]
I hadn't considered the temporal case where A finishes rendering at one point in time, then something happens below A where A needs to know about it.  You raise an interesting case!  Since that is more of an asynchronous and repeatable action, I've seen an event system employed to raise events about what's happening for a component (and children), and this is more akin to a lifecycle like @vvs is saying.  

Re: promises: I'm down with promises and love them a lot.  A nice feature of them vs. callbacks is that you can check a promise after it's been resolved and proceed with the function in then() whereas in the callback case you need to be waiting for the action to happen before it happens.  Using promises also makes sense in the case of wanting to know when a component has been initalized, since that happens once, and you may be looking for it to be finished either before or after it actually initialized.  Callbacks would make sense if you needed repeated notification of things happening inside the component (like event notification of component changes).

**Comment by chrisknoll** [comment]
@vvs: The oracle lifecycle thing (that's actually a knockout implementation too!) looks really really good.  I wonder if you can just drop that into an existing app or if there's some other dependencies on oracle jet that you just can't use that ojModule binding....

**Comment by mbest** [comment]
After some more thought, I'm leaning again toward just having a callback function on the viewmodel. I've created a pull request: #2303

**Comment by mbest** [comment]
I've added #2319, which supports a special callback method on components called `koDescendantsComplete` that is called once all descendant bindings (including async components) are complete.
